src/common/types/row_data_collection_scanner.cpp

```diff
@@ -1,57 +1,193 @@
+//				}
+//				auto next = MinValue<idx_t>(data_block_remaining, heap_block_remaining);
+//
+//				// Figure out where to start copying strings, and how many bytes we need to copy
+//				auto heap_start_ptr = Load<data_ptr_t>(data_ptr + layout.GetHeapOffset());
+//				auto heap_end_ptr =
+//				    Load<data_ptr_t>(data_ptr + layout.GetHeapOffset() + (next - 1) * layout.GetRowWidth());
+//				idx_t size = heap_end_ptr - heap_start_ptr + Load<uint32_t>(heap_end_ptr);
+//				ptrs_and_sizes.emplace_back(heap_start_ptr, size);
+//				D_ASSERT(size <= heap_blocks[heap_block_idx]->byte_offset);
+//
+//				// Swizzle the heap pointer
+//				RowOperations::SwizzleHeapPointer(layout, data_ptr, heap_start_ptr, next, total_size);
+//				total_size += size;
+//
+//				// Update where we are in the data and heap blocks
+//				data_ptr += next * layout.GetRowWidth();
+//				data_block_remaining -= next;
+//				heap_block_remaining -= next;
+//			}
+//
+//			// Finally, we allocate a new heap block and copy data to it
+//			swizzled_string_heap.blocks.emplace_back(
+//			    make_unique<RowDataBlock>(buffer_manager, MaxValue<idx_t>(total_size, (idx_t)Storage::BLOCK_SIZE), 1));
+//			auto new_heap_handle = buffer_manager.Pin(swizzled_string_heap.blocks.back()->block);
+//			auto new_heap_ptr = new_heap_handle.Ptr();
+//			if (swizzled_string_heap.keep_pinned) {
+//				// Since the heap blocks are pinned, we can unswizzle the data again.
+//				swizzled_string_heap.pinned_blocks.emplace_back(move(new_heap_handle));
+//				RowOperations::UnswizzlePointers(layout, base_row_ptr, new_heap_ptr, data_block->count);
+//				RowOperations::UnswizzleHeapPointer(layout, base_row_ptr, new_heap_ptr, data_block->count);
+//			}
+//			for (auto &ptr_and_size : ptrs_and_sizes) {
+//				memcpy(new_heap_ptr, ptr_and_size.first, ptr_and_size.second);
+//				new_heap_ptr += ptr_and_size.second;
+//			}
+//		}
+//	}
+//
+//	// We're done with variable-sized data, now just merge the fixed-size data
+//	swizzled_block_collection.Merge(block_collection);
+//	D_ASSERT(swizzled_block_collection.blocks.size() == swizzled_string_heap.blocks.size());
+//
+//	// Update counts and cleanup
+//	swizzled_string_heap.count = string_heap.count;
+//	string_heap.Clear();
+//}
+
+// Merged with master
+void RowDataCollectionScanner::AlignHeapBlocks(RowDataCollection &dst_block_collection,
+                                               RowDataCollection &dst_string_heap,
+                                               RowDataCollection &src_block_collection,
+                                               RowDataCollection &src_string_heap, const RowLayout &layout) {
+	if (src_block_collection.count == 0) {
 		return;
 	}
 
 	if (layout.AllConstant()) {
 		// No heap blocks! Just merge fixed-size data
-		swizzled_block_collection.Merge(block_collection);
+		dst_block_collection.Merge(src_block_collection);
 		return;
 	}
 
+	// If input data does not have keep_pinned, it should already be aligned
```

Comment:
Is that right? I wanted this to work even when the blocks were not aligned e.g., the `OVER()` case in `PhysicalWindow` where both sets of blocks need to spill. My current long term vision for window is that it will use CDC for parallel hash grouping, then sort the groups, eventually resulting in the current sorted RDCs. Won't that result in unaligned blocks that are swizzled?

Reply:
If `keep_pinned` is false, then the `BufferManager` can decide to evict heap blocks, invalidating pointers. Although I may not be 100% sure about how you're using this. We should have a proper discussion about this at some point.

However, I am leaving for vacation this Friday, and the feature freeze is next week. We want this in before 5.0, so I've decided to postpone fixing this until after I get back. I've now left `AlignHeapBlocks` untouched, and moved my code to `JoinHashTable::SwizzleBlocks`.